harness/determined/common/experimental/checkpoint/_checkpoint.py

```diff
@@ -184,6 +189,53 @@ def download(self, path: Optional[str] = None) -> str:
 
         return str(local_ckpt_dir)
 
+    def _download_direct(
+        self, checkpoint_storage: Dict[str, Any], local_ckpt_dir: pathlib.Path
+    ) -> None:
+        if checkpoint_storage["type"] == "shared_fs":
+            src_ckpt_dir = self._find_shared_fs_path(checkpoint_storage)
+            shutil.copytree(str(src_ckpt_dir), str(local_ckpt_dir))
+        else:
+            local_ckpt_dir.mkdir(parents=True, exist_ok=True)
+            manager = storage.build(
+                checkpoint_storage,
+                container_path=None,
+            )
+            if not isinstance(
+                manager,
+                (
+                    storage.S3StorageManager,
+                    storage.GCSStorageManager,
+                    storage.AzureStorageManager,
+                ),
+            ):
+                raise AssertionError(
+                    "Downloading from Azure, S3 or GCS requires the experiment "
+                    "to be configured with Azure, S3 or GCS checkpointing"
+                    ", {} found instead".format(checkpoint_storage["type"])
+                )
+
+            manager.download(self.uuid, str(local_ckpt_dir))
+
+    @staticmethod
+    def _download_via_master(sess: api.Session, uuid: str, local_ckpt_dir: pathlib.Path) -> None:
+        """Downloads a checkpoint through the master.
+        Arguments:
+            sess (api.Session): a session for the download
+            uuid (string): the uuid of the checkpoint to be downloaded
+            local_ckpt_dir (Path-like): the local directory where the checkpoint is downloaded
+        """
+        local_ckpt_dir.mkdir(parents=True, exist_ok=True)
+
+        resp = sess.get(f"/checkpoints/{uuid}", headers={"Accept": "application/gzip"}, stream=True)
+        if not resp.ok:
+            raise RuntimeError(
+                "unable to download checkpoint from master:", resp.status_code, resp.reason
+            )
+        # gunzip and untar. tarfile.open can detect the compression algorithm
+        with tarfile.open(fileobj=resp.raw) as tf:
+            tf.extractall(local_ckpt_dir)
```

Comment:
It's actually dangerous to call `extractall` without any verification of the paths inside the tarball.

When I was thinking about this yesterday, I was thinking that it was ok to not verify because the master is creating tarballs on the fly so it could guarantee that any tarballs it serves are well-behaved.  But right now, I think it would be possible to craft an s3 bucket with keys like `checkpoints/UUID/../.ssh/config` that could result in the master generating malicious tarballs.

Could we make a ticket to make sure the master validates the paths it puts into the tarball?  There's no reason why a checkpoint should ever contain an `..` element in any path name.

Then we could put a comment here saying that the master validates tarballs before emitting them so this is a fairly trustworthy tarball source.

Reply:
That makes sense and I will file a ticket. I spend too much time on SaaS and keep forgetting a lot of users have direct access to checkpoint storage.

Reply:
https://determinedai.atlassian.net/browse/DET-8501